### ISSUE TITLE
Fine control of fp32 precison in backward phase

### ISSUE THREAD

**Issue by xinyazhang** [pytorch/pytorch#167254]
### 🚀 The feature, motivation and pitch

# Overview

Currently the autograd does not record the FP32 precision used when constructing the graph.
The actual FP32 precision used during the backward phase is completely determined by the configuration of current context.

# Behavior

I added debugging prints at https://github.com/pytorch/pytorch/blob/c5593e75b31286dc5358a451ae6e8c61e47921d6/aten/src/ATen/cuda/CUDABlas.cpp#L401-L403 and https://github.com/pytorch/pytorch/blob/c5593e75b31286dc5358a451ae6e8c61e47921d6/aten/src/ATen/cuda/CUDABlas.cpp#L1619-L1621
and run the following script
```python
#!/usr/bin/env python
import torch

# Set device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Set TF32 precision using new API (PyTorch 2.9+)
torch.backends.cuda.matmul.fp32_precision = "tf32"
torch.backends.cudnn.conv.fp32_precision = "tf32"

def main(bwd_precision):
	# Set random seed for reproducibility
	torch.manual_seed(42)


	# Create random tensors on CUDA with requires_grad=True
	A = torch.randn(3, 4, device=device, dtype=torch.float32, requires_grad=True)
	B = torch.randn(4, 5, device=device, dtype=torch.float32, requires_grad=True)
	C = torch.randn(5, 2, device=device, dtype=torch.float32, requires_grad=True)

	# Matrix multiplication with FP32
	torch.backends.cuda.matmul.fp32_precision = "ieee"
	print("Doing matmul AxB with ieee")
	tmp = torch.matmul(A, B)
	torch.backends.cuda.matmul.fp32_precision = "tf32"
	print("Doing matmul (AxB)xC")
	D = torch.matmul(tmp, C)
	print("Done all matmul")

	# Scalar loss to backpropagate
	loss = D.sum()

	# Compute gradients
	# but default to tf32
	torch.backends.cuda.matmul.fp32_precision = bwd_precision
	print("Doing loss.backward() with global", bwd_precision)
	loss.backward()
	print("Done loss.backward()")
	# Print gradients
	print(bwd_precision, "Gradient w.r.t A:\n", A.grad)
	print(bwd_precision, "Gradient w.r.t B:\n", B.grad)
	print(bwd_precision, "Gradient w.r.t C:\n", C.grad)

main("tf32")
main("ieee")
```

This script prints
```
Doing matmul AxB with ieee                                                                                             
Effective cuda.matmul.fp32_precision = ieee                
Doing matmul (AxB)xC                                       
Effective cuda.matmul.fp32_precision = tf32                
Done all matmul                                            
Doing loss.backward() with global tf32                     
Effective cuda.matmul.fp32_precision = tf32                
Effective cuda.matmul.fp32_precision = tf32                                                                            
Effective cuda.matmul.fp32_precision = tf32                
Effective cuda.matmul.fp32_precision = tf32                
Done loss.backward()
tf32 Gradient w.r.t A:
# Omitted content.
Doing matmul AxB with ieee                                 
Effective cuda.matmul.fp32_precision = ieee                
Doing matmul (AxB)xC                                       
Effective cuda.matmul.fp32_precision = tf32                
Done all matmul                                            
Doing loss.backward() with global ieee                     
Effective cuda.matmul.fp32_precision = ieee                
Effective cuda.matmul.fp32_precision = ieee                
Effective cuda.matmul.fp32_precision = ieee                
Effective cuda.matmul.fp32_precision = ieee                
Done loss.backward()                                       
ieee Gradient w.r.t A:
# Omitted content.
...

Could we add some way to control the fp32 precision in the backward direction so they would follow the forward direction. More concretely in the example above, when calculating the gradients of `tmp = AxB`, "ieee" is used, and "tf32" is used when calculating `D = tmp x C`

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan

**Comment by xinyazhang** [comment]
@meghendra6 I'm thinking of doing this at autograd level. For example the precision can be captured in `MatmulBackward0`'s constructor, and precision context manager can be added in `MatmulBackward0::apply_with_saved`
